Shadowserver 2022-01 update

elsif2 opened the pull request:
This commit is to update the Shadowserver bots to support all current report types.

## Test Results

```
sudo -u intelmq INTELMQ_SKIP_INTERNET=1 python3 -m unittest
----------------------------------------------------------------------
Ran 1457 tests in 74.351s

OK (skipped=246)
```

sebix commented:
I'm not sure if all of this PR was created on purpose. It contains PR#2134 and 6991597278b7286099212e9469f7b097af238adf (not as commit, but as a separate change?), but also tons of other changes in the shadowserver parser which I don't understand and some updates to the shadowserver collector, which need to be reviewed. Many updates to the tests, also new test which I appreciate highly! All in all you changes 142 files in just one commit. Can you please split that enormous commit into smaller ones? E.g. the shadowserver collector, update to the parser and test fixes, new tests or so.

elsif2 committed:
Replaced misleading _country_ parameter with _reports_.

elsif2 committed:
Replaced misleading _country_ parameter with _reports_.

elsif2 committed:
Update parser to support all available reports.

Update to existing test cases to match current report types.

New tests for added report types.

pycodestyle fixes

add testdata licenses

pycodestyle fix

Added reports parameter

Suggested changes to the parser

Proposed details for the release

Test script updates for suggested changes

Test input updates

Realign columns

Update compromised_website.csv

Update scan_adb.csv

Update scan_adb.csv

Update scan_ftp.csv

Update scan_ipp.csv

Update scan_snmp.csv

Realign columns

Remove duplicates

Changed malware.name to extra.infection

Updated SPDX-FileCopyrightText

shadowserver api: document and warn on old parameter

document the old parameter `country` and its status
warn if used
adapt the test

DOC: fix NEWS entry of PR#2143

Added the sector field to scan_amqp, scan_cwmp, and scan_vnc.

Copyright and raw field updates

Added the sector field to scan_amqp, scan_cwmp, and scan_vnc.

Copyright updates

Added phish_url and scan_modbus reports.

Update source.url and source.fqdn for phish_url and malware_url reports.  Update classification.taxonomy and classification.type for scan_modbus report.

* additional field type validation changes
* added count, bytes, duration, avg_pps, and max_pps fields to event_honeypot_ddos_amp
* added 'protocol.application': 'https' to scan_ssl, scan_ssl_freak, and scan_ssl_poodle
* added 'extra.tag' to scan_* and device_id

Replaced scan_modbus with scan_ics

Addeed event4_honeypot_ddos, event4_honeypot_ddos_target, scan_dvr_dhcpdiscover, and scan_socks.

Tests for event4_honeypot_ddos.

Tests for event4_honeypot_ddos_target.

Tests for scan_dvr_dhcpdiscover.

Tests for scan_socks.

Rename file

Rename file

update:scan_mdns, scan_smb, and special; add:scan_ddos_middle_box

cleanup renamed license files

updated scan_mdns test files

updated scan_smb test files

updated special test files

add scan_ddos_middlebox test files

add scan_ddos_middlebox test

updated schema

Updated scan_smb tests

Updated scan_ntp tests

Updated scan_snmp tests

New scan_docker test

New scan_kubernetes test

New scan_mysql test

Updated report schema for June 2022

Added scan_epmd test

Revert "Added scan_epmd test"

This reverts commit 01edea18d82b583290f66eae5f2457f4695623cd.

Revert: Fix for recover_line method as commited in certtools#2192

Added scan_couchdb

Test case for scan_couchdb

Added scan6_rpd

Added/updated README with maintainer details

Restored feed names and classification.identifiers to minimize upgrade impact.

Merge repair

pycodestyle repairs

codespell fixes

license compliance fixes

pycodestyle fixes

Feed configuration updates for compatibility with the original.

Added scan_postgres test

Added additional IPv6 aliases

Fix for recover_line method as commited in certtools#2192

elsif2 committed:
Update parser to support all available reports.

elsif2 committed:
Update to existing test cases to match current report types.

elsif2 committed:
New tests for added report types.

elsif2 committed:
pycodestyle fixes

elsif2 committed:
add testdata licenses

elsif2 committed:
pycodestyle fix

sebix commented on intelmq/bots/parsers/shadowserver/parser.py:
```diff
@@ -86,8 +86,8 @@ def parse_line(self, row, report):
         # https://github.com/certtools/intelmq/issues/1271
-        if conf == config.drone and row.get('infection') == 'spam':
-            conf = config.drone_spam
+        # if conf == config.drone and row.get('infection') == 'spam':
+        #    conf = config.drone_spam
```
Was the feed changed? If yes, please remove the lines entirely.

elsif2 replied:
Removed as drone_spam is no longer supported.

Updated in a2aeda778d7e139eae453d608114b4f1fb017636.

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_broken.py:
```diff
@@ -62,11 +60,11 @@ def test_half_broken(self):
-        self.assertLogMatches(pattern="Optional key 'protocol' not found in feed 'Accessible-FTP'. Possible change in data format or misconfiguration.",
-                              levelname="WARNING")
+        #self.assertLogMatches(pattern="Optional key 'protocol' not found in feed 'Accessible FTP'. Possible change in data format or misconfiguration.",
+        #                      levelname="WARNING")
```
If this is warning is no longer true, please do not comment it out, but replace it with some other assertion

elsif2 replied:
Replaced with a 'jarm' assertion.

Updated in 655055d169257c1f93ee18bda28bdc28600ed5ed

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_device_id.py:
```diff
@@ -0,0 +1,113 @@
+   'extra.device_vendor' : 'Fortinet',
+   'extra.source.naics' : 517311,
+   'feed.name' : 'Device ID',
+   'malware.name' : 'ssl,vpn',
```
`vpn` does not look like a malware to me :thinking:

elsif2 replied:
Fixed:
```
-  'malware.name' : 'ssl,vpn',
+  'extra.tag' : 'ssl,vpn',
```

Updated in 655055d169257c1f93ee18bda28bdc28600ed5ed

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_event4_sinkhole_http_referer.py:
```diff
@@ -23,7 +23,8 @@
            'feed.name': 'Sinkhole-Events-HTTP-Referer IPv4',
-           'classification.identifier': 'sinkhole-http-referer',
+           'classification.identifier': 'event-sinkhole-http-referer',
+           'extra.tag': 'kovter',
```
Is this the malware name?

elsif2 replied:
Fixed:

``` 
-  'extra.tag': 'kovter',
+  'malware.name': 'kovter',
```

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_sandbox_conn.py:
```diff
@@ -0,0 +1,96 @@
+   'classification.identifier' : 'sandbox-conn',
+   'classification.taxonomy' : 'malicious-code',
+   'classification.type' : 'malware-distribution',
+   'extra.host' : 'time.windows.com',
```
is this the destination fqdn? if not, what "host" is it?

elsif2 replied:
Fixed:
```
-    'extra.host' : 'time.windows.com', 
+    'destination.fqdn' : 'time.windows.com',
```

Updated in 655055d169257c1f93ee18bda28bdc28600ed5ed

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_sandbox_dns.py:
```diff
@@ -0,0 +1,92 @@
+   'classification.type' : 'other',
+   'extra.request' : 'time.windows.com',
+   'extra.response' : '40.119.6.228',
+   'extra.type' : 'A',
```
```suggestion
   'extra.dns_query_type' : 'A',
```

elsif2 replied:
Fixed:
```
-   'extra.type' : 'A',
+   'extra.dns_query_type' : 'A',
```
Updated in 655055d169257c1f93ee18bda28bdc28600ed5ed

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_sandbox_url.py:
```diff
@@ -0,0 +1,101 @@
+   'classification.taxonomy' : 'malicious-code',
+   'classification.type' : 'malware-distribution',
+   'extra.host' : 'www.msftncsi.com',
+   'extra.method' : 'GET',
```
```suggestion
   'extra.request_method' : 'GET',
```
As above, make it conform with other parsers

elsif2 replied:
Fixed:
```
-   'extra.method' : 'GET',
+   'extra.http_request_method' : 'GET',
```
Updated in 655055d169257c1f93ee18bda28bdc28600ed5ed

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_sandbox_url.py:
```diff
@@ -0,0 +1,101 @@
+   'classification.identifier' : 'sandbox-url',
+   'classification.taxonomy' : 'malicious-code',
+   'classification.type' : 'malware-distribution',
+   'extra.host' : 'www.msftncsi.com',
```
Same question as above

elsif2 replied:
Fixed:
```
-     'extra.host' : 'www.msftncsi.com',
+     'destination.fqdn' : 'www.msftncsi.com',
```

Updated in 655055d169257c1f93ee18bda28bdc28600ed5ed

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_sandbox_url.py:
```diff
@@ -0,0 +1,101 @@
+   'classification.type' : 'malware-distribution',
+   'extra.host' : 'www.msftncsi.com',
+   'extra.method' : 'GET',
+   'extra.url' : 'http://www.msftncsi.com/ncsi.txt',
```
What URL is this? The requested one?

elsif2 replied:
Fixed:
```
-    'extra.url' : 'http://www.msftncsi.com/ncsi.txt',
+    'destination.url' : 'http://www.msftncsi.com/ncsi.txt',
```

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_scan_synfulknock.py:
```diff
@@ -0,0 +1,114 @@
+   'classification.identifier' : 'scan-synfulknock',
+   'classification.taxonomy' : 'vulnerable',
+   'classification.type' : 'vulnerable-system',
+   'extra.ack_number' : '791102',
```
could this be int?

elsif2 replied:
Fixed:
```
-    'extra.ack_number' : '791102',
+    'extra.ack_number' : 791102,
```

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_scan_synfulknock.py:
```diff
@@ -0,0 +1,114 @@
+   'extra.ack_number' : '791102',
+   'extra.raw_packet' : '3cfdfec601e4700f6a9a2000080045000034c3780000f706789442099555b869f7ee0050b20800000000000c123e8012200002aa0000020405b40101040201030305',
+   'extra.tag' : 'synfulknock',
+   'extra.tcp_flags' : '4608',
+   'extra.window_size' : '8192',
```
could this be int?

elsif2 replied:
Keeping tcp_flags as a string as the implementation could change.

Fixed:
```
-    'extra.window_size' : '8192',
+    'extra.window_size' : 8192,

sebix commented on intelmq/tests/bots/parsers/shadowserver/testdata/compromised_website.csv:
```diff
@@ -1,3 +1,3 @@
-"timestamp","ip","port","hostname","tag","application","asn","geo","region","city","url","http_host","category","system","detected_since","server","redirect_target","naics","sic","sector"
+"timestamp","ip","port","hostname","tag","application","asn","geo","region","city","url","http_host","category","system","detected_since","server","redirect_target","naics","sic","sector","cc_url","family"
```
are these two columns missing in the data rows?

elsif2 replied:
Fixed.

sebix commented on intelmq/tests/bots/parsers/shadowserver/testdata/scan_adb.csv:
```diff
@@ -1,3 +1,3 @@
-"timestamp","ip","protocol","port","hostname","tag","asn","geo","region","city","naics","sic","name","model","device","features"
+"timestamp","ip","protocol","port","hostname","tag","asn","geo","region","city","naics","sic","name","model","device","features","device_vendor","device_type","device_model","device_version","device_sector"
```
as above: looks like the data is missing

elsif2 replied:
Fixed

sebix commented on intelmq/tests/bots/parsers/shadowserver/testdata/scan_coap.csv:
```diff
@@ -1,2 +1,2 @@
-"timestamp","ip","protocol","port","hostname","tag","asn","geo","region","city","naics","sic","response"
+"timestamp","ip","protocol","port","hostname","tag","asn","geo","region","city","naics","sic","response","version"
```
missing data?

elsif2 replied:
Fixed.

sebix commented on intelmq/tests/bots/parsers/shadowserver/testdata/scan_ftp.csv:
```diff
@@ -1,3 +1,3 @@
-"timestamp","ip","protocol","port","hostname","tag","asn","geo","region","city","naics","sic","banner","handshake","cipher_suite","cert_length","subject_common_name","issuer_common_name","cert_issue_date","cert_expiration_date","sha1_fingerprint","cert_serial_number","ssl_version","signature_algorithm","key_algorithm","subject_organization_name","subject_organization_unit_name","subject_country", ...
+"timestamp","ip","protocol","port","hostname","tag","asn","geo","region","city","naics","sic","banner","handshake","cipher_suite","cert_length","subject_common_name","issuer_common_name","cert_issue_date","cert_expiration_date","sha1_fingerprint","cert_serial_number","ssl_version","signature_algorithm","key_algorithm","subject_organization_name","subject_organization_unit_name","subject_country", ...
```
missing data?

elsif2 replied:
Fixed.

sebix commented on intelmq/tests/bots/parsers/shadowserver/testdata/scan_ipp.csv:
```diff
@@ -1,2 +1,2 @@
-"timestamp","ip","protocol","port","hostname","tag","asn","geo","region","city","naics","sic","ipp_version","cups_version","printer_uris","printer_name","printer_info","printer_more_info","printer_make_and_model","printer_firmware_name","printer_firmware_string_version","printer_firmware_version","printer_organization","printer_organization_unit","printer_uuid","printer_wifi_ssid"
+"timestamp","ip","protocol","port","hostname","tag","asn","geo","region","city","naics","sic","ipp_version","cups_version","printer_uris","printer_name","printer_info","printer_more_info","printer_make_and_model","printer_firmware_name","printer_firmware_string_version","printer_firmware_version","printer_organization","printer_organization_unit","printer_uuid","printer_wifi_ssid","device_vendor" ...
 "2020-06-08 11:30:14","123.45.67.89","tcp",631,"some.host.com","ipp",12345,"AA","REGION","CITY",517311,0,"IPP/2.1","CUPS/2.0","ipp://123.45.67.89:631/ipp/print","NPI3F0D22","HP Color LaserJet MFP M277dw","http://123.45.67.89:631/hp/device/info_config_AirPrint.html?tab=Networking&menu=AirPrintStatus","HP Color LaserJet MFP M277dw",20191203,20191203,20191203,"org","unit","urn:uuid:456e4238-4a44-464 ...
```
missing data?

elsif2 replied:
Fixed.

sebix commented on intelmq/tests/bots/parsers/shadowserver/testdata/scan_mqtt.csv:
```diff
@@ -1,3 +1,3 @@
-"timestamp","ip","protocol","port","hostname","tag","asn","geo","region","city","naics","sic","anonymous_access","raw_response","hex_code","code"
+"timestamp","ip","protocol","port","hostname","tag","asn","geo","region","city","naics","sic","anonymous_access","raw_response","hex_code","code","cert_expiration_date","cert_issue_date","cert_length","cert_serial_number","cipher_suite","issuer_business_category","issuer_common_name","issuer_country","issuer_email_address","issuer_given_name","issuer_locality_name","issuer_organization_name","iss ...
 "2020-03-14 05:45:48","123.45.67.89","tcp",1883,"some.host.com","mqtt",12345,"AA","REGION","CITY",518210,0,"Y",20020000,00,"Connection Accepted"
 "2020-03-14 05:45:51","123.45.67.90","tcp",1883,"another.host.com","mqtt",12345,"AA","REGION","CITY",454110,0,"N",20020003,03,"Connection Refused, Server unavailable"
```
missing data?

elsif2 replied:
Fixed.

sebix commented on intelmq/tests/bots/parsers/shadowserver/testdata/scan_snmp.csv:
```diff
@@ -1,4 +1,4 @@
-"timestamp","ip","protocol","port","hostname","sysdesc","sysname","asn","geo","region","city","version","naics","sic","sector"
+"timestamp","ip","protocol","port","hostname","sysdesc","sysname","asn","geo","region","city","version","naics","sic","sector","device_model","device_sector","device_type","device_vendor","device_version"
 "2014-03-16 03:45:50","129.113.21.93","udp",161,"doesnotexist.utpa.edu","Hardware: x86 Family 6 Model 8 Stepping 6 AT/AT COMPATIBLE - Software: Windows 2000 Version 5.0 (Build 2195 Uniprocessor Free)","ORSONKA",22864,"US","TEXAS","EDINBURG",2,0,0,
 "2014-03-16 03:45:51","79.2.242.16","udp",17080,"host16-242-dynamic.2-79-r.retail.telecomitalia.it","ADSL Modem","tc",3269,"IT","EMILIA-ROMAGNA","RAVENNA",2,0,0,
 "2014-03-16 03:45:51","95.109.21.127","udp",161,"ip6-127.skekraft.riksnet.se",,,34610,"SE","VASTERBOTTENS LAN","UMEA",2,0,0,
```
missing data?

elsif2 replied:
Fixed.

sebix commented on intelmq/tests/bots/parsers/shadowserver/testdata/scan_ssdp.csv:
```diff
@@ -1,3 +1,3 @@
-"timestamp","ip","protocol","port","hostname","tag","header","asn","geo","region","city","systime","cache_control","location","server","search_target","unique_service_name","host","nts","nt","naics","sic","sector"
+"timestamp","ip","protocol","port","hostname","tag","header","asn","geo","region","city","systime","cache_control","location","server","search_target","unique_service_name","host","nts","nt","naics","sic","sector","content_type","instance","resource_identifier","server_port","updated_at","version"
 "2019-09-04 07:13:42","198.123.245.171","udp",32822,,"ssdp","HTTP/1.1 200 OK",5678,"AA","LOCATION","LOCATION","Fri, 16 Jan 1970 00:05:23 GMT","max-age=1800","http://198.123.245.1:52869/gatedesc.xml","Linux, UPnP/1.0, Portable SDK for UPnP devices/1.6.6","upnp:rootdevice","uuid:20809696-105a-3721-e8b8-2c957f6259fe::upnp:rootdevice",,,,517311,0,
 "2019-09-04 07:13:50","198.123.245.237","udp",32818,"localhost.localdomain","ssdp","HTTP/1.1 200 OK",5678,"AA","LOCATION","LOCATION","Thu, 20 Jan 2000 12:14:46 GMT","max-age=1800","http://198.123.245.1:52869/gatedesc.xml","Linux/2.6.20-Amazon_SE, UPnP/1.0, Intel SDK for UPnP devices /1.2","upnp:rootdevice","uuid:160a0200-ac91-4b05-8adf-f5ccc5a5ebaa::upnp:rootdevice",,,,517311,0,
```
missing data?

elsif2 replied:
Fixed.

sebix commented on intelmq/tests/bots/parsers/shadowserver/testdata/scan_ssl_freak.csv:
```diff
@@ -1,4 +1,4 @@
-"timestamp","ip","port","hostname","tag","handshake","asn","geo","region","city","cipher_suite","cert_length","subject_common_name","issuer_common_name","cert_issue_date","cert_expiration_date","sha1_fingerprint","cert_serial_number","signature_algorithm","key_algorithm","subject_organization_name","subject_organization_unit_name","subject_country","subject_state_or_province_name","subject_locali ...
+"timestamp","ip","port","hostname","tag","handshake","asn","geo","region","city","cipher_suite","cert_length","subject_common_name","issuer_common_name","cert_issue_date","cert_expiration_date","sha1_fingerprint","cert_serial_number","signature_algorithm","key_algorithm","subject_organization_name","subject_organization_unit_name","subject_country","subject_state_or_province_name","subject_locali ...
 "2018-04-23 13:25:21","198.51.100.232","443",,"ssl-freak","TLSv1.0","8447","AT","WIEN","VIENNA","TLS_RSA_WITH_RC4_128_SHA","1024","usg50_B0B2DC2FA69D","usg50_B0B2DC2FA69D","2012-05-10 00:01:19","2032-05-05 00:01:19","14:09:8C:6E:64:5F:50:C9:E9:A3:62:5E:02:BB:33:67:E1:05:D3:D2","4FAB054F","sha1WithRSAEncryption","rsaEncryption",,,,,,,,,,,,,,,,,,,,,,,,,"0","0","Y","TLS_RSA_EXPORT_WITH_RC4_40_MD5",, ...
 "2018-04-23 13:25:26","198.51.100.224","443","198-51-100-224.example.net","ssl-freak","TLSv1.0","12577","AT","NIEDEROSTERREICH","BADEN","TLS_RSA_WITH_RC4_128_SHA","1024","usg20w_C86C870287EC","usg20w_C86C870287EC","2010-01-01 00:00:53","2029-12-27 00:00:53","14:09:8C:6E:64:5F:50:C9:E9:A3:62:5E:02:BB:33:67:E1:05:D3:D2","4B3D3B35","sha1WithRSAEncryption","rsaEncryption",,,,,,,,,,,,,,,,,,,,,,,,,"0", ...
 2018-04-23 13:25:21,198.51.100.232,443,,ssl-freak,TLSv1.0,8447,AT,WIEN,VIENNA,TLS_RSA_WITH_RC4_128_SHA,1024,usg50_B0B2DC2FA69D,usg50_B0B2DC2FA69D,2012-05-10 00:01:19,2032-05-05 00:01:19,14:09:8C:6E:64:5F:50:C9:E9:A3:62:5E:02:BB:33:67:E1:05:D3:D2,4FAB054F,sha1WithRSAEncryption,rsaEncryption,,,,,,,,,,,,,,,,,,,,,,,,,0,0,Y,TLS_RSA_EXPORT_WITH_RC4_40_MD5,,57:7A:FC:7C:A1:0F:79:11:67:E0:31:AC:66:F5:84:2 ...
```
missing data?

elsif2 replied:
Fixed.

sebix commented on intelmq/tests/bots/parsers/shadowserver/testdata/scan_ssl_poodle.csv:
```diff
@@ -1,4 +1,4 @@
-"timestamp","ip","port","hostname","tag","handshake","asn","geo","region","city","cipher_suite","ssl_poodle","cert_length","subject_common_name","issuer_common_name","cert_issue_date","cert_expiration_date","sha1_fingerprint","cert_serial_number","ssl_version","signature_algorithm","key_algorithm","subject_organization_name","subject_organization_unit_name","subject_country","subject_state_or_pro ...
+"timestamp","ip","port","hostname","tag","handshake","asn","geo","region","city","cipher_suite","ssl_poodle","cert_length","subject_common_name","issuer_common_name","cert_issue_date","cert_expiration_date","sha1_fingerprint","cert_serial_number","ssl_version","signature_algorithm","key_algorithm","subject_organization_name","subject_organization_unit_name","subject_country","subject_state_or_pro ...
 "2018-08-08 00:51:42","203.0.113.85",8443,"example.com","ssl-poodle","TLSv1.0",65540,"AT","WIEN","VIENNA","TLS_RSA_WITH_RC4_128_SHA","Y",1024,"usg20_107BEF394BA5","usg20_107BEF394BA5","2014-06-25 00:00:42","2034-06-20 00:00:42","04:FA:DE:1D:BD:4A:05:25:61:FB:F3:D6:64:74:66:44:01:22:D7:C3","53AA112A",2,"sha1WithRSAEncryption","rsaEncryption",,,,,,,,,,,,,,,,,,,,,,,,,0,0,,"16:25:9F:C7:A1:8D:64:1F:D9 ...
 2018-04-19 13:32:27,198.51.100.147,443,,ssl-poodle,TLSv1.0,8445,AT,SALZBURG,SALZBURG,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,Y,2048,198-51-100-147.example.net,some_issuer,2017-09-18 08:22:17,2019-09-18 08:22:17,14:09:8C:6E:64:5F:50:C9:E9:A3:62:5E:02:BB:33:67:E1:05:D3:D2,746481F100000000000C,2,sha1WithRSAEncryption,rsaEncryption,some_org_name,some_org_name,AT,Tirol,Ehrwald,,,,,,,,,,,,,,,,,,,,0,0,,57:7A ...
 2018-04-19 13:32:32,198.51.100.207,443,198-51-100-94.example.net,ssl-poodle,TLSv1.0,25255,AT,SALZBURG,SALZBURG,TLS_RSA_WITH_RC4_128_SHA,Y,1024,example,some_issuer,2004-06-03 11:11:43,2024-05-29 11:11:43,14:09:8C:6E:64:5F:50:C9:E9:A3:62:5E:02:BB:33:67:E1:05:D3:D2,0,2,md5WithRSAEncryption,rsaEncryption,,,US,,,,,,,,,,,,US,,,,,,,,,,0,0,,57:7A:FC:7C:A1:0F:79:11:67:E0:31:AC:66:F5:84:22:28:4E:AC:9D:27:A ...
```
missing data?

elsif2 replied:
Fixed.

sebix commented on intelmq/bots/collectors/shadowserver/collector_reports_api.py:
```diff
@@ -36,6 +39,8 @@ class ShadowServerAPICollectorBot(CollectorBot, HttpMixin, CacheMixin):
     api_key = None
     secret = None
     types = None
+    reports = None
```
Please describe this parameter in docs/user/bots.rst

elsif2 replied:
Added.

sebix commented on intelmq/bots/collectors/shadowserver/collector_reports_api.py:
```diff
@@ -36,6 +39,8 @@ class ShadowServerAPICollectorBot(CollectorBot, HttpMixin, CacheMixin):
     secret = None
     types = None
+    reports = None
+    report_list = []
```
```suggestion
    _report_list = []
```
Looks like it is an internal variable?

elsif2 replied:
Fixed:
```
-      report_list = []
+      _report_list = []
```

sebix commented on intelmq/bots/collectors/shadowserver/collector_reports_api.py:
```diff
@@ -48,11 +53,12 @@ def init(self):
 
-        if isinstance(self.types, str):
-            self.types = self.types.split(',')
+        if isinstance(self.reports, str):
+            self.report_list = self.reports.split(',')
```
What if self.reports is not a string, but e.g. a list? self.report_list stays None?

elsif2 replied:
Fixed:
```
        if isinstance(self.reports, str):
            self._report_list = self.reports.split(',')
        elif isinstance(self.reports, list):
            self._report_list = self.reports
```

sebix reviewed (changes requested):
First of all, thank you for this PR, mus have been a lot of work!

There lots of changes which I am unable to review properly. Someone with access to the data and the API needs to cover these parts.
Maybe @monoidic or @waldbauer-certat ?

There are a few changes which should be documented in the NEWS file:
- change of the identifier of the blocklist, honeypot feeds, ip spoofer, MS sinkhole, ..... That will cause troubles for the users and should therefore be documented very well, e.g. with a table
  - And I don't see a good reason to change "open" and "vulnerable" to "scan". "Scan" describes they way how the information was retrieved, but not what the event is about. The IoCs do not describe, that the address do scanning, but that they are vulnerable.
- change of `extra.naics` -> `extra.source.naics`

From the formal side a summary of the changes is necessary to communicate to the users what they need to adapt in their workflows.

elsif2 committed:
update internal variable name and related list issue

elsif2 committed:
Added reports parameter

elsif2 committed:
Suggested changes to the parser

elsif2 committed:
Proposed details for the release

elsif2 committed:
Test script updates for suggested changes

elsif2 committed:
Test input updates

elsif2 committed:
Update compromised_website.csv

elsif2 committed:
Update scan_adb.csv

elsif2 committed:
Update scan_adb.csv

elsif2 committed:
Update scan_ftp.csv

elsif2 committed:
Update scan_ipp.csv

elsif2 committed:
Update scan_snmp.csv

elsif2 committed:
Realign columns

elsif2 committed:
Merge branch 'elsif-2022-01' of https://github.com/elsif2/intelmq into elsif-2022-01

elsif2 committed:
Realign columns

elsif2 committed:
Remove duplicates

elsif2 commented:
I made a number of changes to address your comments.  Please let me know if further changes are required.

aaronkaplan commented:
I have been trying to go through this massive PR and tbh I got interrupted frequently due to work. It's a bit ... huge.
@elsif2 could we maybe arrange for a call and you walk me through this a bit?
I would really like to finally merge this.

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_event4_ip_spoofer.py:
```diff
@@ -36,12 +36,13 @@
         "extra.source.naics": 517311,
         "extra.version": 'ipv4',
         "protocol.transport": 'tcp',
+        "malware.name": 'ip-spoofer',
```
Is `ip-spoofer` really the name of a malware?

elsif2 replied:
Added a special case to the code generator for ip-spoofer.

```
- "malware.name": 'ip-spoofer',
+ "extra.infection": 'ip-spoofer',
```

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_sandbox_conn.py:
```diff
@@ -0,0 +1,96 @@
+# SPDX-FileCopyrightText: 2019 Guillermo Rodriguez
```
2019? This looks like a new file.

elsif2 replied:
Updated copyright.

sebix replied:
This also applies to lots of other files

elsif2 replied:
The copyright has been updated on files added.

sebix committed:
shadowserver api: document and warn on old parameter

document the old parameter `country` and its status
warn if used
adapt the test

sebix committed:
DOC: fix NEWS entry of PR#2143

sebix reviewed (changes requested):
I applied some fixes on the NEWS and added a deprecation warning to the collector.

I leave it for someone else to write a changelog and fix the merge conflict. Before merging, the commits need to be squashed.

sebix commented:
> There lots of changes which I am unable to review properly. Someone with access to the data and the API needs to cover these parts.
> Maybe @monoidic or @waldbauer-certat ?

And that still applies.

elsif2 committed:
Changed malware.name to extra.infection

elsif2 committed:
Updated SPDX-FileCopyrightText

elsif2 committed:
Merge branch 'elsif-2022-01' of https://github.com/elsif2/intelmq into elsif-2022-01

elsif2 committed:
Added the sector field to scan_amqp, scan_cwmp, and scan_vnc.

elsif2 committed:
Copyright and raw field updates

elsif2 committed:
Added the sector field to scan_amqp, scan_cwmp, and scan_vnc.

elsif2 committed:
Copyright updates

elsif2 committed:
Added phish_url and scan_modbus reports.

sebix commented on intelmq/bots/parsers/shadowserver/_config.py:
```diff
@@ -223,1375 +223,1407 @@ def force_base64(value: Optional[str]) -> Optional[str]:
-        ('extra.', 'version', validate_to_none),
-        ('source.asn', 'asn', invalidate_zero),
+        ('extra.', 'url', validate_to_none),
+        ('extra.', 'host', validate_to_none),
```
If that's the URL/host of the phishing page:
```suggestion
        ('source.url', 'url', validate_to_none),
        ('source.fqdn', 'host', validate_to_none),
```

elsif2 replied:
Updated:

```
-        ('extra.', 'url', validate_to_none),
-        ('extra.', 'host', validate_to_none),
+        ('source.url', 'url', convert_http_host_and_url, True),
+        ('source.fqdn', 'host', validate_fqdn),
```

sebix commented on intelmq/bots/parsers/shadowserver/_config.py:
```diff
@@ -1605,1461 +1637,1599 @@ def force_base64(value: Optional[str]) -> Optional[str]:
-    }
+        'classification.taxonomy': 'other',
+        'protocol.application': 'modbus',
+        'classification.type': 'other',
```
Why not `vulnerable-system`?

elsif2 replied:
Updated:

```
         'protocol.application': 'modbus',
-        'classification.taxonomy': 'other',
-        'classification.type': 'other',
+        'classification.taxonomy': 'vulnerable',
+        'classification.type': 'vulnerable-system',

```

sebix commented on intelmq/tests/bots/parsers/shadowserver/test_malware_url.py:
```diff
@@ -0,0 +1,110 @@
+   'classification.type' : 'malware-distribution',
+   'extra.application' : 'http',
+   'extra.host' : '41.86.0.0',
+   'extra.url' : 'http://41.86.0.0:50008/Mozi.m',
```
```suggestion
   'source.ip' : '41.86.0.0',
   'source.url' : 'http://41.86.0.0:50008/Mozi.m',
```

elsif2 replied:
Updated:

```
-        ('extra.', 'url', validate_to_none),
-        ('extra.', 'host', validate_to_none),
+        ('source.url', 'url', convert_http_host_and_url, True),
+        ('source.fqdn', 'host', validate_fqdn),
```

elsif2 committed:
Update source.url and source.fqdn for phish_url and malware_url reports.  Update classification.taxonomy and classification.type for scan_modbus report.

elsif2 committed:
* additional field type validation changes
* added count, bytes, duration, avg_pps, and max_pps fields to event_honeypot_ddos_amp
* added 'protocol.application': 'https' to scan_ssl, scan_ssl_freak, and scan_ssl_poodle
* added 'extra.tag' to scan_* and device_id

elsif2 commented:
Here is a list of the reports where the _classification.identifier_ changed: 
[classification.identifier.txt](https://github.com/certtools/intelmq/files/8165788/classification.identifier.txt)


Here is a report by report diff for `bots/parsers/shadowserver/_config.py`:
[_config.py.diff.txt](https://github.com/certtools/intelmq/files/8165792/_config.py.diff.txt)

elsif2 committed:
Replaced scan_modbus with scan_ics

monoidic commented:
> > There lots of changes which I am unable to review properly. Someone with access to the data and the API needs to cover these parts.
> > Maybe @monoidic or @waldbauer-certat ?
> 
> And that still applies.

I do not currently have access to Shadowserver and will, in general, be unavailable in 2022.

elsif2 commented:
I have added a couple new features to the API to make integration testing easier:

## Sample report content is available with the `test` reports filter:

```
python3 call-api.py reports/list '{"reports":["test"],"limit":3}' pretty

[
   {
      "report" : "test",
      "timestamp" : "2010-02-10",
      "file" : "2010-02-10-scan_elasticsearch-test.csv",
      "type" : "scan_elasticsearch",
      "id" : "test:scan_elasticsearch"
   },
   {
      "report" : "test",
      "timestamp" : "2010-02-10",
      "file" : "2010-02-10-scan_exchange-test.csv",
      "type" : "scan_exchange",
      "id" : "test:scan_exchange"
   },
   {
      "report" : "test",
      "timestamp" : "2010-02-10",
      "file" : "2010-02-10-blocklist-test.csv",
      "id" : "test:blocklist",
      "type" : "blocklist"
   }
]
```

This provides a list of all possible report types with sample downloads.

## Sample report content is available by supplying an id with a `test:` prefix followed by the report type:

```
python3 call-api.py reports/download '{"id":"test:scan_vnc", "limit":3}' pretty
[
   {
      "banner" : "RFB 003.003|kVNC Server is not licensed correctly.||Connections will be prohibited until a valid license key is applied.",
      "geo" : "ZZ",
      "ip" : "192.168.0.1",
      "naics" : 0,
      "region" : "Region",
      "sector" : "",
      "hostname" : "node01.example.com",
      "asn" : 64512,
      "port" : "5900",
      "sic" : 0,
      "city" : "City",
      "timestamp" : "2010-02-10 00:00:00",
      "product" : "kVNC (Unlicensed)"
   },
   {
      "banner" : "RFB 005.000",
      "geo" : "ZZ",
      "ip" : "192.168.0.2",
      "naics" : 0,
      "region" : "Region",
      "sector" : "Communications, Service Provider, and Hosting Service",
      "hostname" : "node02.example.com",
      "asn" : 64512,
      "port" : "5900",
      "sic" : 0,
      "city" : "City",
      "timestamp" : "2010-02-10 00:00:01",
      "product" : "RealVNC Enterprise v5.3 or later"
   },
   {
      "banner" : "RFB 003.003|Too many security failures",
      "geo" : "ZZ",
      "ip" : "192.168.0.3",
      "naics" : 0,
      "region" : "Region",
      "sector" : "Communications, Service Provider, and Hosting Service",
      "hostname" : "node03.example.com",
      "asn" : 64512,
      "port" : "5901",
      "sic" : 0,
      "city" : "City",
      "timestamp" : "2010-02-10 00:00:02",
      "product" : "VNC protocol 3.3 (locked out)"
   }
]
```

aaronkaplan commented:
very very cool. Thanks!



> On 09.03.2022, at 23:45, elsif2 ***@***.***> wrote:
> 
> 
> I have added a couple new features to the API to make integration testing easier:
> 
> Sample report content is available with the test reports filter:
> 
> python3 call-api.py reports/list '{"reports":["test"],"limit":3}' pretty
> 
> [
>    {
>       "report" : "test",
>       "timestamp" : "2010-02-10",
>       "file" : "2010-02-10-scan_elasticsearch-test.csv",
>       "type" : "scan_elasticsearch",
>       "id" : "test:scan_elasticsearch"
>    },
>    {
>       "report" : "test",
>       "timestamp" : "2010-02-10",
>       "file" : "2010-02-10-scan_exchange-test.csv",
>       "type" : "scan_exchange",
>       "id" : "test:scan_exchange"
>    },
>    {
>       "report" : "test",
>       "timestamp" : "2010-02-10",
>       "file" : "2010-02-10-blocklist-test.csv",
>       "id" : "test:blocklist",
>       "type" : "blocklist"
>    }
> ]
> 
> This provides a list of all possible report types with sample downloads.
> 
> Sample report content is available by supplying an id with a test: prefix followed by the report type:
> 
> python3 call-api.py reports/download '{"id":"test:scan_vnc", "limit":3}' pretty
> [
>    {
>       "banner" : "RFB 003.003|kVNC Server is not licensed correctly.||Connections will be prohibited until a valid license key is applied.",
>       "geo" : "ZZ",
>       "ip" : "192.168.0.1",
>       "naics" : 0,
>       "region" : "Region",
>       "sector" : "",
>       "hostname" : "node01.example.com",
>       "asn" : 64512,
>       "port" : "5900",
>       "sic" : 0,
>       "city" : "City",
>       "timestamp" : "2010-02-10 00:00:00",
>       "product" : "kVNC (Unlicensed)"
>    },
>    {
>       "banner" : "RFB 005.000",
>       "geo" : "ZZ",
>       "ip" : "192.168.0.2",
>       "naics" : 0,
>       "region" : "Region",
>       "sector" : "Communications, Service Provider, and Hosting Service",
>       "hostname" : "node02.example.com",
>       "asn" : 64512,
>       "port" : "5900",
>       "sic" : 0,
>       "city" : "City",
>       "timestamp" : "2010-02-10 00:00:01",
>       "product" : "RealVNC Enterprise v5.3 or later"
>    },
>    {
>       "banner" : "RFB 003.003|Too many security failures",
>       "geo" : "ZZ",
>       "ip" : "192.168.0.3",
>       "naics" : 0,
>       "region" : "Region",
>       "sector" : "Communications, Service Provider, and Hosting Service",
>       "hostname" : "node03.example.com",
>       "asn" : 64512,
>       "port" : "5901",
>       "sic" : 0,
>       "city" : "City",
>       "timestamp" : "2010-02-10 00:00:02",
>       "product" : "VNC protocol 3.3 (locked out)"
>    }
> ]
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.
> You are receiving this because you commented.
>

elsif2 committed:
Addeed event4_honeypot_ddos, event4_honeypot_ddos_target, scan_dvr_dhcpdiscover, and scan_socks.

elsif2 committed:
Tests for event4_honeypot_ddos.

elsif2 committed:
Tests for event4_honeypot_ddos_target.

elsif2 committed:
Tests for scan_dvr_dhcpdiscover.

elsif2 committed:
Tests for scan_socks.

elsif2 committed:
Rename file

elsif2 committed:
Rename file

elsif2 committed:
update:scan_mdns, scan_smb, and special; add:scan_ddos_middle_box

elsif2 committed:
cleanup renamed license files

elsif2 committed:
updated scan_mdns test files

elsif2 committed:
updated scan_smb test files

elsif2 committed:
updated special test files

elsif2 committed:
add scan_ddos_middlebox test files

elsif2 committed:
add scan_ddos_middlebox test

elsif2 committed:
updated schema

elsif2 committed:
Updated scan_smb tests

elsif2 committed:
Updated scan_ntp tests

elsif2 committed:
Updated scan_snmp tests

elsif2 committed:
New scan_docker test

elsif2 committed:
New scan_kubernetes test

elsif2 committed:
New scan_mysql test

sebix commented:
The amount of changes here is so immense that a review in this form is practically impossible. I can image that reviewing and picking commit by commit could be an option to tackle this huge stack of changes, but the re-organization (sorting?) of the `_config.py` in e6a0714f5c808973b7fb7f05e13c35c372328531 renders even that approach painstaking.
I have no good idea how to proceed here.

aaronkaplan commented:
So concerning @sebix 's comment above "The amount of changes here is so immense that a review in this form is practically impossible. [...]", I talked with Shadowserver (Piotr) these days. The summary was: 1) it's a bit too hard for us to review 2) but @elsif2 will take the responsibility to fix something if it breaks. So, I will - as agreed with Shadowserver - merge this PR and hand over the responsibility for maintainership for shadowerver bots to them. 
And actually that is a much better approach.

@elsif2 let's have a chat how we can make ongoing , strict but useful, unit tests for all of them. And we will be good in the future.
I will merge this independently of the unit test discussion ok?

elsif2 commented:
> @elsif2 let's have a chat how we can make ongoing , strict but useful, unit tests for all of them. And we will be good in the future. I will merge this independently of the unit test discussion ok?

Each report type has a unit test following the exiting format.  If something more comprehensive is needed, please provide a template and I will generate code that follows it.

sebix commented:
@aaronkaplan Thanks for this information. If @elsif2 takes over the maintenance, that would indeed relief all other users and devs.
Please also document this in the header of the affected source code files.

Please don't forget that this PR's code is not compatible with current IntelMQ configurations (and I see no apparent reason for these changes) so users will need to adapt here, interrupting the automatic upgrade procedure. To prevent this, I propose to add an upgrade method to `intelmq/lib/upgrades.py` to automatically change the configuration, reducing not only the frustration on the users side, but also to minimize the errors when doing so.
And an entry to the changelog is missing.

elsif2 committed:
Updated report schema for June 2022

elsif2 committed:
Added scan_epmd test

elsif2 committed:
Fix for recover_line method as commited in #2192

elsif2 committed:
Revert "Added scan_epmd test"

This reverts commit 01edea18d82b583290f66eae5f2457f4695623cd.

elsif2 committed:
Resync

elsif2 committed:
Revert: Fix for recover_line method as commited in certtools#2192

elsif2 committed:
Added scan_couchdb

elsif2 committed:
Test case for scan_couchdb

elsif2 committed:
Added scan6_rpd

elsif2 committed:
Added/updated README with maintainer details

elsif2 commented:
> Please don't forget that this PR's code is not compatible with current IntelMQ configurations (and I see no apparent reason for these changes) so users will need to adapt here, interrupting the automatic upgrade procedure. To prevent this, I propose to add an upgrade method to `intelmq/lib/upgrades.py` to automatically change the configuration, reducing not only the frustration on the users side, but also to minimize the errors when doing so. And an entry to the changelog is missing.

Are the incompatible configurations the changes to just the _classification.identifier_ or does it include changes to the _classification.type_ also?

sebix commented:
> > Please don't forget that this PR's code is not compatible with current IntelMQ configurations (and I see no apparent reason for these changes) so users will need to adapt here, interrupting the automatic upgrade procedure. To prevent this, I propose to add an upgrade method to `intelmq/lib/upgrades.py` to automatically change the configuration, reducing not only the frustration on the users side, but also to minimize the errors when doing so. And an entry to the changelog is missing.
> 
> Are the incompatible configurations the changes to just the _classification.identifier_ or does it include changes to the _classification.type_ also?

You are mixing two topics here. One is the runtime configuration of IntelMQ. You changed the first column of `mapping`, which is the feed name, an (optional) parameter of the shadowserver parser. This can be handled by an upgrade function so that the users do not need to adapt this parameter by hand.

The other topic is a bigger issue. You have changed various mapped values, which means the resulting, parsed data is then different to what it was before. Until now, we have been very hesitant to do this, as it causes lots of work for the users. They need to check and adapt all processing steps after the shadowserver bots. Not only the bot's configurations (e.g. filtering, rules, notifications, ...), but also any further processing scripts outside the IntelMQ Core (and many users have a lot of them) and including changes to their databases to have consistent values for statistics etc. That's a huge effort for everyone. To at least aid in this last step, we have provided UPDATE-statements in the past. Just have a look a the NEWS file, you'll find some examples there.

elsif2 commented:
> The other topic is a bigger issue. You have changed various mapped values, which means the resulting, parsed data is then different to what it was before. Until now, we have been very hesitant to do this, as it causes lots of work for the users. They need to check and adapt all processing steps after the shadowserver bots. Not only the bot's configurations (e.g. filtering, rules, notifications, ...), but also any further processing scripts outside the IntelMQ Core (and many users have a lot of them) and including changes to their databases to have consistent values for statistics etc. That's a huge effort for everyone. To at least aid in this last step, we have provided UPDATE-statements in the past. Just have a look a the NEWS file, you'll find some examples there.

I am working on an revision that will set the _classification.identifier_ to match the current `develop` branch.

Once that is done, the only substantial change to the mapped values impacts the configurations where _extra.naics_, _extra.sector_, and _extra.sic_ were used instead of _extra.source.naics_, _extra.source.sector_, and _extra.source.sic_.   This change should make it easier for users as the fields are consistent across all report types.

I suppose another option would be to configure both the non-source and source versions for those three fields.  Please let me know your preference.

elsif2 committed:
Restored feed names and classification.identifiers to minimize upgrade impact.

elsif2 committed:
Added function to notify if any unsupported feeds are configured.

elsif2 commented:
I was not sure what release version to use in `intelmq/lib/upgrades.py` - currently set for 3.1.0.

I think the only outstanding issue is how to handle the source.naics/sic/sector changes.

elsif2 committed:
resolve conflict

elsif2 committed:
Merge branch 'develop' into elsif-2022-01

elsif2 committed:
pycodestyle repairs

elsif2 committed:
codespell fixes

elsif2 committed:
license compliance fixes

elsif2 committed:
Merge branch 'elsif-2022-01' of https://github.com/elsif2/intelmq into elsif-2022-01

elsif2 committed:
pycodestyle fixes

elsif2 committed:
Feed configuration updates for compatibility with the original.

elsif2 committed:
Added scan_postgres test

elsif2 commented:
I completed the compatibility updates.  Here is a diff showing the changes for each feed:
[diff.txt](https://github.com/certtools/intelmq/files/9096087/diff.txt)

I believe this PR is ready to be merged.  Please let me know if you have any suggestions or concerns.

sebix commented:
According to @aaronkaplan, @elsif2 himself is responsible for this PR.

elsif2 committed:
Merge conflict correction

elsif2 committed:
Fix conflicts

elsif2 commented:
> According to @aaronkaplan, @elsif2 himself is responsible for this PR.

I am not authorized to merge a pull request.  I created a clean branch for this PR as [shadowserver-202207](https://github.com/certtools/intelmq/tree/shadowserver-202207).  Please advise on how you would like me to proceed.

sebix commented:
> > According to @aaronkaplan, @elsif2 himself is responsible for this PR.
> 
> I am not authorized to merge a pull request.

I'm caught between two stools, as I don't know the details of your agreement with @aaronkaplan. Can you please check with him?

I'd like to do a release of IntelMQ (3.1.0), and the shadowserver updates would be the major feature of the new release.

elsif2 commented:
Replaced by #2227
